README.md
# SLALite #

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![CircleCI](https://circleci.com/gh/mF2C/SlaManagement.svg?style=svg)](https://circleci.com/gh/mF2C/SlaManagement)

## Description ##

The SLALite is a lightweight implementation of an SLA system, inspired by the
WS-Agreement standard. Its features are:

* REST interface to manage creation and update of agreements
* Agreements evaluation on background; any breach in the agreement terms
  generates an SLA violation.
* Configurable monitoring: a monitoring has to be provided externally.
* Configurable repository: a memory repository (for developing purposes)
  and a mongodb repository are provided, but more can be added.

An agreement is represented by a simple JSON structure 
(see examples in resources/samples):

```
{
    "id": "2018-000234",
    "name": "an-agreement-name",
    "details":{
        "id": "2018-000234",
        "type": "agreement",
        "name": "an-agreement-name",
        "provider": { "id": "a-provider", "name": "A provider" },
        "client": { "id": "a-client", "name": "A client" },
        "creation": "2018-01-16T17:09:45Z",
        "expiration": "2019-01-17T17:09:45Z",
        "guarantees": [
            {
                "name": "TestGuarantee",
                "constraint": "[execution_time] < 100"
            }
        ]
    }
}
```

## Quick usage guide ##

### Installation ###

Build the Docker image:

    make docker

Run the container:

    docker run -ti -p 8090:8090 slalite:<version>

Stop execution pressing CTRL-C

To run the service under HTTPs, you must change supply a different configuration file and the certificate files. You will find these files in docker/https for debugging purposes. DO NOT USE THE CERT.PEM and KEY.PEM in production!!

    docker run -ti -p 8090:8090 -v $PWD/docker/https:/etc/slalite slalite

### Configuration ###

The SLALite can be configured with a configuration file and with environment 
variables. The configuration file is read by default from /etc/slalite and the current 
working directory. The `-f` parameter can be used to set the config file location.

```
$ ./SLALite -h
Usage of SLALite:
  -b string
        Filename (w/o extension) of config file (default "slalite")
  -d string
        Directories where to search config files (default "/etc/slalite:.")
  -f string
        Path of configuration file. Overrides -b and -d
```

#### File settings ####

*General settings*

* `singlefile` (default: `false`). Sets if all file settings are read 
  from a single file or from several files. For example, when `singlefile=false`,
  the MongoDB settings are read from the file `mongodb.yml`.
* `repository` (default: `memory`). Sets the repository type to use. Set this
  value to `mongodb` to use a MongoDB database.
* `externalIDs` (default: `false`). Set this to true if the repository auto assign 
  the IDs of the saved entities.
* `checkPeriod` (default: `60`). Sets the period in seconds of assessments 
  executions.
* `CAPath`. Sets the value of a file path containing certificates of trusted
  CAs; to be used to connect as client to SSL servers whose certificate is
  not trusted by default (e.g. self-signed certificates)

*REST interface settings*

* `port` (default: `8090`). Port of REST interface.
* `enableSsl` (default: `false`). Enables the use of SSL on the REST
  interface. The two following variables should be set.
* `sslCertPath` (default: `cert.pem`). Sets the certificate path.
* `sslKeyPath` (default: `key.pem`). Sets the private key path to access the
  certificate.

*MongoDB settings (default file: /etc/slalite/mongodb.yml)*

* `connection` (default: `localhost`). Sets the MongoDB host.
* `database` (default: `slalite`). Sets the MongoDB database name to use.
* `clear_on_boot` (default: `false`). Sets if the database is cleared on
  startup (useful for tests).

*mF2C settings*

The recommended way of running the mF2C SLA Management is using the Docker image 
(https://hub.docker.com/r/mf2c/sla-management) in the docker-compose file
(https://github.com/mF2C/mF2C).

The parameters below and default values are intended for development purposes.

* `policies` (default: `http://localhost:46050/api/v2`). Sets the root location of the Policies component API.
* `isleader` (default: ""). If not empty, this setting, interpreted as a boolean, is used to check if the SLALite is running on a leader or not.
* `analytics` (default: `http://localhost:46020`). Sets the root location of the Analytics component API.

#### Env vars  ####

Every file setting can be overriden with the use of environment variables.
The name of the var is the uppercase setting name prefixed with `SLA_`. For
example, to override the check period, set the env var `SLA_CHECKPERIOD`.

### Usage ###

SLALite offers a usual REST API, with an endpoint on /agreements

Add an agreement (agreement below is stopped):

    curl -k -X POST -d @resources/samples/agreement.json http://localhost:8090/agreements

Change agreement state:

    curl -k http://localhost:8090/agreements/a02 -X PUT -d'{"state":"started"}'

Get agreements:

    curl -k http://localhost:8090/agreements
    curl -k http://localhost:8090/agreements/a02

Add a template:

    curl -k -X POST -d @resources/samples/template.json http://localhost:8090/templates

Get templates:

    curl -k http://localhost:8090/templates
    curl -k http://localhost:8090/templates/t01

Create agreement from template:

    curl -k -X POST -d @resources/samples/create-agreement.json http://localhost:8090/create-agreement

    {"template_id":"t01","agreement_id":"9be511e8-347f-4a40-b784-e80789e4c65b","parameters":{"M":1,"N":100,"agreementname":"An agreement name","client":{"id":"client01","name":"A name of a client"},"provider":{"id":"provider01","name":"A name of a provider"}}}

#### mF2C ####

The following steps are suitable for the default settings for the mF2C project 
(i.e., running with CIMI as repository)

Add a template:

    OUT=$(curl -k -X POST -d @resources/samples/template01cimi.json http://localhost:46030/templates)
    TEMPLATE_ID=$(echo "$OUT" | jq -r ".id")

Create agreement from template. The needed property is the user launching the service:

    IN=$(<resources/samples/create-agreement-cimi.json jq ".template_id=\"$TEMPLATE_ID\"")
    OUT=$(echo "$IN" | curl -k -X POST -d @- http://localhost:46030/mf2c/create-agreement)

Get agreement created:

    AGREEMENT_ID=$(echo $OUT | jq -r ".agreement_id" | sed -e 's/agreement\///')
    curl http://localhost:46030/agreements/$AGREEMENT_ID


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

## v0.9.1 - 2020-01-13

* Fix service operation reports being ignored [Román SG]

## v0.9.0 - 2019-10-24

* Add availability calculation for COMPSs services (#28) [Román SG]
* Update README.md with circleci and license badges [Román SG]
* Feature/circleci (#27) [Román SG]

## v0.8.1 - 2019-08-09

* Fix assessm panic due to CIMI deleting empty maps [Román SG]

## v0.8.0 - 2019-07-19

* Add current example of availability and usage [Román SG]
* Add integration test for new CIMI adapter [Román SG]
* Support availability in SLAs [Román SG]
* Update swagger definitions [Román SG]
* Add scope, schedule; and refactor assessment model [Román SG]
* Use generic adapter and random samples in Main [Román SG]
* Add Assessment.LastValues to agreement model [Román SG]
* Change model of Details.Variables to an array [Román SG]
* Add RandomRetrieve, holistic test and bug fixes [Román SG]
* Add generic adapter and most of the tests [Román SG]
* Add Variable to model [Román SG]
* Make MonitoringAdapter API stateless [Román SG]

## v0.7.2 - 2019-06-05

* Trigger patch [Román SG]
* Adapt to changes in ServiceOperationReport [Román SG]
* Add repo tests for ServiceOperationReport [Román SG]
* Add repo support for ServiceContainerMetric [Román SG]

## v0.7.1 - 2019-05-27

* Update swagger definition file [Román SG]
* Add `release` target to Makefile [Román SG]

## v0.7.0 - 2019-05-24

* Add versionist to manage changelog.md [Román SG]
* Add analytics url to docker config file [Román SG]
* Reintroduce createViolation on cimi test [Román SG]
* Fix cyclic dependency error in cimi tests [Román SG]
* Add generate agreement from template docs [Román SG]
* Add support for Analytics component [Román SG]
* Prepare mf2c package to add more components [Román SG]

## v0.6.0 - 2019-04-17

* Adapt to policies:2.0 version [Román SG]
* Fix agreement validation error on externalId repo [Román SG]
* Add mF2C generate agreement from template and test [Román SG]
* Make CIMI integration tests pass [Román SG]
* Add REST to create agreements from templates [Román SG]
* Clean cache (build, tests...) on make clean [Román SG]
* Add externalIDs config variable [Román SG]
* Rename Validater interface to Validator [Román SG]
* Move repository_testbase.go to pkg repositories [Román SG]
* Improve repositories.validation coverage [Román SG]
* Apply visitor pattern to model validation [Román SG]
* Update swagger file with template resource [Román SG]
* Fix failing test [Román SG]
* Add rest of Template endpoints and tests [Román SG]
* Add templates endpoints [Román SG]
* Add templates methods to repository and memrepo [Román SG]
* Simplify repository tests [Román SG]
* Improve generator comments [Román SG]
* Add template validation; increase model coverage [Román SG]
* Move code of utils.ReadTemplate,ReadAgreement to model [Román SG]
* First draft of agreement generation [Román SG]

## v0.5.0 - 2019-02-12

* Add call to Policies component in assessment process [Román SG]
* Modify main.go code to create Policies component and pass to assessment [Román SG]
* Add REST client to Policies component [Román SG]
* Add base REST client code to build REST clients [Román SG]
* Fix fatal call with formatting directive [Román SG]
* Add swagger annotations [José Antonio Sánchez Murillo]
* Improve Dockerfile [Román SG]
* Fix compilation error on previous commit [Román SG]

## v0.4.0 - 2018-12-03

* Add CIMI repo integration test [Román SG]
* Allow modifying the testing entities from derived tests [Román SG]
* Allow modifying the testing entities from derived tests [Román SG]
* Add (integration) test for mongodb repository [Román SG]
* Add generic tests for repositories [Román SG]
* Add update agreement and terminate agreement REST operations [Román SG]
* Fix utils/ssl_test failing because of expired certificate [Román SG]
* Include timestamps in violation values [Jose Antonio Sanchez]
* Add omitempty for non expirable slas [Jose Antonio Sanchez]
* Simplify monitoring workflow [Jose Antonio Sanchez]
* Make expiration date optional [Jose Antonio Sanchez]
* Fix replace GetActiveAgreements with GetAgreementsByState [Román SG]
* Fix missing tests for assessment.model.Result.getViolations [Román SG]
* Use dep for dependency management [Jose Antonio Sanchez]

## v0.3.1 - 2018-06-20

* Fix wrong default value of cimiurl var [Román SG]
* Update README with CAPath variable [Román SG]
* Allow adding additional trusted CAs to connect to SSL servers [Román SG]

## v0.3.0 - 2018-06-06

* Fix issue 18 : add "catch all" guarantee term [Román SG]
* Externalize creation of docker image to a script [Román SG]
* Update README with configuration instructions [Román SG]
* Accept CIMI_URL as env var to set CIMI location [Román SG]
* Change all references to logging to logrus Add x-example for parameters in swagger [Jose Antonio Sanchez]
* Fix compilation and run with memory repository by default [Jose Antonio Sanchez]
* Use logrus as log library in main and assessment [Román SG]

## v0.2.2 - 2018-05-18

* Add translation from model.Violation to cimimodel.Violation [Román SG]
* Fix Docker image name in Makefile [Román SG]

## v0.2.1 - 2018-05-17

* Add log for every REST path simplifying configuration [Román SG]
* Assessment fills new Violation fields [Román SG]

## v0.2.0 - 2018-05-15

* Fix tests to support added violation fields [Román SG]
* Add constraint and values fields to violation entity [Román SG]
* Fix tests [Román SG]
* Fix adapter initialized twice [Román SG]
* Add simple versioning mechanism to SLALite [Román SG]
* Fix #12: Wrong ID to look for ServiceOperationReport [Román SG]
* Fix #13, add delete agreement and fix bug accesing resource by id [Román SG]
* Support external ids repository on validation decorator [Román SG]

model/model.go
/*
Copyright 2017 Atos

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/*
Package model contain the entities used in the SLALite: agreements, violations, penalties...
It also defines the interface IRepository, which defines the operations to be implemented
by any repository.
*/
package model

import (
	"errors"
	"fmt"
	"time"
)

//
// ErrNotFound is the sentinel error for an entity not found
//
var ErrNotFound = errors.New("Entity not found")

//
// ErrAlreadyExist is the sentinel error for creating an entity whose id already exists
//
var ErrAlreadyExist = errors.New("Entity already exists")

/*
 * ValidationErrors following behavioral errors
 * (https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully)
 */

// validationError is an interface that must be implemented by custom error implementations
// swagger:ignore
type validationError interface {
	IsErrValidation() bool
}

//
// IsErrValidation return true is an error is a validation error
//
func IsErrValidation(err error) bool {
	v, ok := err.(validationError)
	return ok && v.IsErrValidation()
}

// func IsErrNotFound(err error) bool

//
// Identity identifies entities with an Id field
//
type Identity interface {
	GetId() string
}

//
// Validable identifies entities that can be validated
//
type Validable interface {
	Validate(val Validator, mode ValidationMode) []error
}

// State is the type of possible states of an agreement
type State string

// TextType is the type of possible types a Details type
type TextType string

// AggregationType is the type of supported variable aggregations
type AggregationType string

const (
	// STARTED is the state of an agreement that can be evaluated
	STARTED State = "started"

	// STOPPED is the state of an agreement temporaryly not evaluated
	STOPPED State = "stopped"

	// TERMINATED is the final state of an agreement
	TERMINATED State = "terminated"
)

const (
	// AGREEMENT is the text type of an Agreement text
	AGREEMENT TextType = "agreement"

	// TEMPLATE is the text type of a Template text
	TEMPLATE TextType = "template"
)

const (
	// NONE is used when the variable is not aggregated
	NONE AggregationType = "none"
	// AVERAGE is used to calculate average of a variable
	AVERAGE AggregationType = "average"
)

// States is the list of possible states of an agreement/template
var States = [...]State{STOPPED, STARTED, TERMINATED}

// Party is the entity that represents a service provider or a client
// swagger:model
type Party struct {
	Id   string `json:"id" bson:"_id"`
	Name string `json:"name"`
}

// Provider is the entity that represents a Provider
// swagger:model
type Provider Party

// GetId returns the Id of a provider
func (p *Provider) GetId() string {
	return p.Id
}

// Validate validates the consistency of a Provider entity
func (p *Provider) Validate(val Validator, mode ValidationMode) []error {
	return val.ValidateProvider(p, mode)
}

// Client is the entity that represents a client.
// swagger:model
type Client Party

// GetId returns the Id of a client
func (c *Client) GetId() string {
	return c.Id
}

// Validate validates the consistency of a Client entity
func (c *Client) Validate(val Validator, mode ValidationMode) []error {
	return val.ValidateClient(c, mode)
}

// Template is the entity that serves as base to create new agreements
//
// The Details field of the template contains placeholders that are substituted
// when generating an agreement from a template (see generator package).
// The Constraints fields contains constraints that a variable used in a guarantee
// must satisfy. F.e., if the guarantee expression is "cpu_usage < {{M}}", one could
// specify in Constraints that "M" : "M >= 0 && M <= 100".Template
//
// The Id and Name are relative to the template itself, and should not match
// the fields in Details.
// swagger:model
type Template struct {
	Id      string  `json:"id" bson:"_id"`
	Name    string  `json:"name"`
	State   State   `json:"state"`
	Details Details `json:"details"`
	// mF2C commented Constraints map[string]string `json:"constraints"`
}

// CreateAgreement is the resource used to create an agreement from a template.
// swagger:model
type CreateAgreement struct {
	TemplateID  string                 `json:"template_id"`
	AgreementID string                 `json:"agreement_id"`
	Parameters  map[string]interface{} `json:"parameters"`
}

// Agreement is the entity that represents an agreement between a provider and a client.
// The Text is ReadOnly in normal conditions, with the exception of a renegotiation.
// The Assessment cannot be modified externally.
// The Signature is the Text digitally signed by the Client (not used yet)
// swagger:model
type Agreement struct {
	Id         string      `json:"id" bson:"_id"`
	Name       string      `json:"name"`
	State      State       `json:"state"`
	Assessment *Assessment `json:"assessment,omitempty"`
	Details    Details     `json:"details"`

	/* Signature string `json:"signature"` */
}

// Assessment is the struct that provides assessment information
// swagger:model
type Assessment struct {
	FirstExecution time.Time `json:"first_execution,omitempty"`
	LastExecution  time.Time `json:"last_execution,omitempty"`
	// Guarantees may be nil. Use Assessment.SetGuarantee to create if needed.
	Guarantees map[string]AssessmentGuarantee `json:"guarantees,omitempty"`
}

// AssessmentGuarantee contain the assessment information for a guarantee term
//
// swagger:model
type AssessmentGuarantee struct {
	FirstExecution time.Time  `json:"first_execution"`
	LastExecution  time.Time  `json:"last_execution"`
	LastValues     LastValues `json:"last_values,omitempty"`
}

// LastValues contain last values of variables in guarantee terms
//
// swagger:model
type LastValues map[string]MetricValue

// Details is the struct that represents the "contract" signed by the client
// swagger:model
type Details struct {
	Id         string      `json:"id"`
	Type       TextType    `json:"type"`
	Name       string      `json:"name"`
	Provider   Provider    `json:"provider"`
	Client     Client      `json:"client"`
	Creation   time.Time   `json:"creation"`
	Expiration *time.Time  `json:"expiration,omitempty"`
	Variables  []Variable  `json:"variables,omitempty"`
	Guarantees []Guarantee `json:"guarantees"`
}

// Variable gives additional information about a metric used in a Guarantee constraint
// swagger:model
type Variable struct {
	Name        string       `json:"name"`
	Metric      string       `json:"metric"`
	Aggregation *Aggregation `json:"aggregation,omitempty"`
}

// Aggregation gives aggregation information of a variable.
// If defined and value is not NONE, the metric must be aggregated
// in the specified window in seconds.
// I.e. (average, 3600) means that the average over a period of one hour is calculated.
// swagger:model
type Aggregation struct {
	Type   AggregationType `json:"type"`
	Window int             `json:"window"`
}

// Guarantee is the struct that represents an SLO
// swagger:model
type Guarantee struct {
	Name       string       `json:"name"`
	Scope      Scope        `json:"scope,omitempty"`
	Constraint string       `json:"constraint"`
	Schedule   Schedule     `json:"schedule,omitempty"`
	Warning    string       `json:"warning,omitempty"`
	Penalties  []PenaltyDef `json:"penalties,omitempty"`
}

// Scope is the resources a guarantee term applies on
type Scope string

// Schedule is the frequency a guarantee term is evaluated
type Schedule string

// PenaltyDef is the struct that represents a penalty in case of an SLO violation
// swagger:model
type PenaltyDef struct {
	Type  string `json:"type"`
	Value string `json:"value"`
	Unit  string `json:"unit"`
}

// MetricValue is the SLALite representation of a metric value.
// swagger:model
type MetricValue struct {
	Key      string      `json:"key"`
	Value    interface{} `json:"value"`
	DateTime time.Time   `json:"datetime"`
}

func (v MetricValue) String() string {
	return fmt.Sprintf("{Key: %s, Value: %v, DateTime: %v}", v.Key, v.Value, v.DateTime)
}

// Violation is generated when a guarantee term is not fulfilled
// swagger:model
type Violation struct {
	Id          string        `json:"id"`
	AgreementId string        `json:"agreement_id"`
	Guarantee   string        `json:"guarantee"`
	Datetime    time.Time     `json:"datetime"`
	Constraint  string        `json:"constraint"`
	Values      []MetricValue `json:"values"`
}

// Penalty is generated when a guarantee term is violated is the term has
// PenaltyDefs associated.
// swagger:model
type Penalty struct {
	Id          string     `json:"id"`
	AgreementId string     `json:"agreement_id"`
	Guarantee   string     `json:"guarantee"`
	Datetime    time.Time  `json:"datetime"`
	Definition  PenaltyDef `json:"definition"`
}

// GetId returns the id of an template
func (t *Template) GetId() string {
	return t.Id
}

// Validate validates the consistency of a Template.
func (t *Template) Validate(val Validator, mode ValidationMode) []error {
	return val.ValidateTemplate(t, mode)
}

// GetId returns the id of a CreateAgreement entity
func (ca *CreateAgreement) GetId() string {
	return ""
}

// GetId returns the id of an agreement
func (a *Agreement) GetId() string {
	return a.Id
}

// IsStarted is true if the agreement state is STARTED
func (a *Agreement) IsStarted() bool {
	return a.State == STARTED
}

// IsTerminated is true if the agreement state is TERMINATED
func (a *Agreement) IsTerminated() bool {
	return a.State == TERMINATED
}

// IsStopped is true if the agreement state is STOPPED
func (a *Agreement) IsStopped() bool {
	return a.State == STOPPED
}

// IsValidTransition returns if the transition to newState is valid
func (a *Agreement) IsValidTransition(newState State) bool {
	return a.State != TERMINATED
}

// Validate validates the consistency of an Agreement.
func (a *Agreement) Validate(val Validator, mode ValidationMode) []error {
	return val.ValidateAgreement(a, mode)
}

// Validate validates the consistency of an Assessment entity
func (as *Assessment) Validate(val Validator, mode ValidationMode) []error {
	return val.ValidateAssessment(as, mode)
}

// SetGuarantee is a helper function to set the assessment info of a guarantee term
func (as *Assessment) SetGuarantee(name string, value AssessmentGuarantee) {
	if as.Guarantees == nil {
		as.Guarantees = make(map[string]AssessmentGuarantee)
	}
	as.Guarantees[name] = value
}

// GetGuarantee is a helper to return the assessment info of a guarantee term.
//
// If empty, it returns a zero AssessmentGuarantee
func (as *Assessment) GetGuarantee(name string) AssessmentGuarantee {
	zero := AssessmentGuarantee{
		LastValues: LastValues{},
	}
	if as.Guarantees == nil {
		return zero
	}

	var ag AssessmentGuarantee
	var ok bool
	if ag, ok = as.Guarantees[name]; !ok {
		return zero
	}
	if ag.LastValues == nil {
		ag.LastValues = LastValues{}
	}
	return ag
}

// Validate validates the consistency of a Details entity
func (t *Details) Validate(val Validator, mode ValidationMode) []error {
	return val.ValidateDetails(t, mode)
}

// GetVariable returns the variable with name "varname".
//
// If not found, it returns a default value for the variable
// (i.e., Name and Metric equal to varname).
func (t *Details) GetVariable(varname string) (result Variable, ok bool) {
	for _, val := range t.Variables {
		if varname == val.Name {
			return val, true
		}
	}
	return Variable{Name: varname, Metric: varname}, false
}

// Validate validates the consistency of a Guarantee entity
func (g *Guarantee) Validate(val Validator, mode ValidationMode) []error {
	return val.ValidateGuarantee(g, mode)
}

// GetId returns the Id of a violation
func (v *Violation) GetId() string {
	return v.Id
}

// Validate validates the consistency of a Violation entity
func (v *Violation) Validate(val Validator, mode ValidationMode) []error {
	return val.ValidateViolation(v, mode)
}

// Normalize returns an always valid state: any different value from contained in States is STOPPED.
func (s State) Normalize() State {
	return normalizeState(s)
}

// Providers is the type of an slice of Provider
// swagger:model
type Providers []Provider

// Agreements is the type of an slice of Agreement
// swagger:model
type Agreements []Agreement

// Templates is the type of an slice of Template
// swagger:model
type Templates []Template
